[MRG] Nested seq speed

darcymason opened the pull request:
Related to one problem in #1728 - accessing Sequences by index caused a big slowdown due to repeated setting of parents down the entire nested sequence.

This PR changes the parent concept:
- before, a dataset's `parent` was set to the 'grandparent' dataset (not recording the sequence the dataset belonged to).  Here there is a `parent_seq` which in turn has a `parent_dataset`.
- setting of weakref's to the parents is done when items are added to the Dataset or Sequence, not on `Dataset.__getitem__`  
- `__deepcopy__`s have been added to Sequence and Dataset to reset the weakrefs after the copy has been done

darcymason commented:
Also added timing tests for these using airspeed velocity.  In the process, moved the benchmarks directory to what seems to be the usual recommended place for it - and added the .json config file for this branch.  I have researched a bit about the ability run asv on a PR by labeling with a `run_benchmark` label, and hopefully can set that up in a separate PR.

darcymason commented:
I keep having thoughts about this -- one problem with changing things this way is that we could have items in a sequence assigned to different parent datasets.  It's unlikely to be a normal use case, but it is possible if someone is (shallow) copying things or setting sequence items in a new dataset/sequence based on cherry-picking from an old one, e.g. a file which has been read in.  The dataset reference would still exist in the old dataset sequence, but it's parent_seq would be the one it was most recently added to.  We could check if it already has a valid parent and warn/raise if so to prevent this.

But ... maybe there are other ways.  I'll take a look at where parents are used, and see if the same functionality cannot be just passed down from the parents so are available when needed; we already mostly pass around `is_little_endian` and `is_implicit_VR` and the character encoding.  Perhaps others can be passed too.

Another thing is perhaps we should put back the old `parent` properties (as deprecated, for removal in 3.0).  It is unlikely they were used in user code, but it wouldn't be hard to add them back in.

darcymason commented:
> I'll take a look at where parents are used, and see if the same functionality cannot be just passed down from the parents so are available when needed;

As far as I can tell on global search, these are only used in `filewriter._correct_ambiguous_vr_element()`, to check PixelRepresentation to differentiate between `US` and `SS` pixels.  The function is passed the DataElement, Dataset, and is_little_endian flag.  It checks for PixelRepresentation in the dataset, but then chains up the parents if it is not found.

Otherwise these `parent...`s appear in the pickling-related code, but that's only necessary because of the weakref's.

As to `correct_ambiguous_vr_element()`, it is called (if needed) in `Dataset.__getitem__` and from `correct_ambiguous_vr()` (used by `write_dataset`) which recurses through a given dataset to correct all of them.  It looks like it will recurse again when each sequence item hits `write_dataset` ... - the recursive correction should happen in `dcmwrite`, I think, so it only happens once.

Lots of details above, just kind of 'talking out loud' as I work through this.

So, we could get rid of the parents concept completely, if we could count on ambiguous data not being nested (but it appears that we did hit cases in the past where it was), or if we did not correct the VR in `Dataset.__getitem__` and only in the recursive function `correct_ambigous_vr()` which can be called from the top level with a `PixelRepresentation` parameter passed down.

I'd suggest we go ahead with this PR's changes, so we have a snapshot of this system, then in a subsequent PR I can try to remove parents completely (and deprecate them in case used in user code).

darcymason commented:
@mrbean-bremen, would appreciate if you would have a quick look at this while I tidy up last remaining errors and typing/style issues.  I'll have to write some tests as well, but wanted to make sure that the switch in ways of tracking parents is conceptually okay.

mrbean-bremen commented:
Sorry, I missed your comment - will have a look tonight or tomorrow.

mrbean-bremen commented on asv.conf.json:
```diff
@@ -0,0 +1,70 @@
+    //    "some_benchmark": 0.01,     // Threshold of 1%
+    //    "another_benchmark": 0.5,   // Threshold of 50%
+    // },
+}
```
Good to have that! I only just installed asv, somehow never got to that... 
Is there a possibilty to run this in CI?

mrbean-bremen commented on pydicom/dataset.py:
```diff
@@ -408,6 +412,73 @@ def __init__(self, *args: _DatasetType, **kwargs: Any) -> None:
+            raise AttributeError("Future: Dataset.parent is removed in v3.x")
+        else:
+            warnings.warn(
+                "Dataset.parent will be removed in pydicom 3.0",
```
This does not do anything - if somebody is indeed using this (which I doubt), this will break the code. I would add the setter code, or at least add a note to the warning that the code is non-functional

darcymason replied:
Adding the actual setting code...

mrbean-bremen commented on pydicom/sequence.py:
```diff
@@ -3,13 +3,17 @@
+from pydicom import config
 from pydicom.dataset import Dataset
 from pydicom.multival import MultiValue
+from pydicom.tag import Tag, TagType, TagListType
```
Is there a reason for these, or is it just a leftover?

mrbean-bremen commented on pydicom/sequence.py:
```diff
@@ -93,32 +107,57 @@ def insert(    # type: ignore[override]
         """
-        if value != self._parent:
-            self._parent = weakref.ref(value)
-            for item in self._list:
```
So I guess this was a performance problem...

darcymason replied:
Yes, when sequences accessed through an index, this was called on every index.  Even if accessed only once, it is a 'costly' operation time-wise for heavily nested large lists.

mrbean-bremen reviewed (commented):
As far as I can see, this looks good, but I will have another look with a bit more time. At the moment, I'm a bit short on time.
Anyway, there are still the failing tests and linter issues to fix...

darcymason commented:
Thanks for the review, @mrbean-bremen - I'll hopefully post corrections today or tomorrow.  We'll still have to decide whether the parent concept is needed at all if we can replace the one place in code (AFAICT) where it is actually used now.

mrbean-bremen commented:
> We'll still have to decide whether the parent concept is needed at all if we can replace the one place in code (AFAICT) where it is actually used now.

This was a fix for the problem of ambiguous VRs in sequences (#804), which needed the access to some top-level tags while decoding these. I do remember that [I wasn't happy](https://github.com/pydicom/pydicom/issues/804#issuecomment-464497220) with the solution, but didn't see another. If we can do this a better way, I'm all for it, of course!

darcymason commented:
@mrbean-bremen, I've added the fixes, and added something new -- checking for internal attributes like is_implicit_VR, encoding, parent things, etc. at the start of `Dataset.__setattr__`, rather than always going through checking if a DICOM keyword, repeater keyword, etc., which was adding a lot of time for all the Datasets in heavily nested sequences.  The new checks should be very fast if setting an actual DICOM keyword, so I don't think should impact old code noticeably.

I'll post some timing results too ... and will work on  adding an asv PR when I get a chance - although I've read an article that github actions timing is probably only good to detect ~50% changes.  Running asv on my laptop was probs good only to ~25%.  Sigh, timing things is always difficult.

darcymason commented:
![asv iterate seq](https://user-images.githubusercontent.com/61110/205713001-f0ac242d-a9aa-4738-b7ec-4f79acad951a.png)

darcymason commented:
![asv iterate preloaded tags](https://user-images.githubusercontent.com/61110/205713149-6ffd57cf-1e07-42e3-b5a8-b32c474e21f1.png)

darcymason commented:
And finally, the really big one (log scale) when indexing was used - going from 10 seconds to tens of ms.
![asv indexing seq](https://user-images.githubusercontent.com/61110/205713521-481daad8-c448-47fd-b3eb-093bf20420aa.png)

mrbean-bremen reviewed (approved):
Looks good to me!

mrbean-bremen commented:
> added something new -- checking for internal attributes like is_implicit_VR, encoding, parent things, etc. at the start of Dataset.__setattr__, rather than always going through checking if a DICOM keyword, repeater keyword,

Yes, good point! Optimizing is fun 😄 

> And finally, the really big one (log scale) when indexing was used - going from 10 seconds to tens of ms.

Well, that shows perfectly well that we shall make performance test after each change. That change looked so innocent at the time... 

> although I've read an article that github actions timing is probably only good to detect ~50% changes.

Well, you can't have it all... even for 50% it is worth it, I think.

darcymason commented on pydicom/dataset.py:
```diff
@@ -2120,6 +2186,16 @@ def __setattr__(self, name: str, value: Any) -> None:
         """
+        # Save time for common Dataset attributes that are not DICOM keywords
+        # This check is fast if `name` is a DICOM keyword (first chr is upper)
+        # The startswith is needed for `is_implicit_VR`
```
That would help with the `islower` thing, however, having `startswith` is probs faster for the common `is_...` case as it only needs to check the first few characters rather than all of them (for those that are indeed not a DICOM keyword).  We could go really fast and just check if the first character is lower case, but then we would lose warnings/errors for camel case like `ds.patientName`.

darcymason commented on pydicom/tests/test_dataset.py:
```diff
@@ -2193,6 +2199,8 @@ def contains_warn():
     config.INVALID_KEY_BEHAVIOR = original
+
+
 CAMEL_CASE = (
```
It is strange, but some (e.g. Rows, BitsStored) are actually DICOM keywords.

darcymason commented on benchmarks/bench_nested_seq.py:
```diff
@@ -0,0 +1,80 @@
+from pydicom.tag import Tag
+
+def create_nested_test_seq(num_items: int = 6280) -> Dataset:
+    """Create a simplified version of sequence from issue #1728"""
```
Ah, yes, forgot to run pycodestyle locally.  I think they are fixed now.
